src/lambdas/fusionauth-site-origin-handler/src/index.js

```diff
@@ -0,0 +1,163 @@
+import { GetObjectCommand, S3Client } from "@aws-sdk/client-s3";
+const s3 = new S3Client({ region: 'us-east-1' });
+
+// Location of the file containing the redirect rules.
+const fileBucket = 'fusionauth-dev-us-east-1-artifacts';
+const fileKey = 'lambda/fusionauth-site-origin-handler/redirects.json';
+
+let redirectRules = null;
+
+// When is the last time the rules were updated? This is stored
+// in globals so it can be shared between lambda invocations.
+let lastUpdatedTime = 0;
+// How frequently (in seconds) should we refresh the redirect rules?
+let intervalBetweenUpdates = 300;
+
+export const handler = async (event, context) => {
+  var req = event.Records[0].cf.request;
+  var uri = req.uri;
+  var result = req;
+
+  await syncRedirectRules()
+
+  try {
+    // Try to load the redirect rules from s3.
+    if (!redirectRules) {
+      console.error("[ERROR] missing redirect rules");
+      return;
+    }
+    // If we successfully retrieved the rules, parse them as JSON.
+    else {
+      redirectRules = JSON.parse(redirectRules);
+    }
+
+    // If any requests come in with .html extensions, trim the extension.
+    if (uri.endsWith('.html')) {
+      result = redir(uri.substring(0, uri.length - 5));
+    }
+    // If this looks like a reference to a file but with a trailing slash, remove it.
+    else if (uri.endsWith('/') && removeSlash(uri)) {
+      result = redir(uri.substring(0, uri.length - 1));
+    }
+    // If this is a reference to an index page that's missing its slash, add it.
+    else if (!uri.endsWith('/') && redirectRules.indexPaths[uri + '/'] === true) {
+      result = redir(uri + '/');
+    }
+    else {
+      var redirect = calculateRedirect(uri);
+      if (redirect !== null) {
+        result = redir(redirect);
+      } else {
+        req.uri = calculateURI(uri);
+      }
+    }
+
+    return result;
+  }
+  catch (err) {
+    console.error('[ERROR] %s', err);
+  }
+};
+
+async function syncRedirectRules() {
+  const command = new GetObjectCommand({
```

Comment:
Might as well move this down below the time check block, since it will be executed on every request and we'll only need it if the time check fails.